File: packages/main-api/src/types/file.rs
use bdk::prelude::*;

#[derive(Debug, serde::Serialize, serde::Deserialize, Clone, JsonSchema, PartialEq, Eq)]
pub struct File {
    #[serde(default = "generate_file_id")]
    pub id: String,
    pub name: String,
    pub size: String,
    pub ext: FileExtension,
    pub url: Option<String>,
}

fn generate_file_id() -> String {
    uuid::Uuid::new_v4().to_string()
}

impl Default for File {
    fn default() -> Self {
        File {
            id: generate_file_id(),
            name: String::new(),
            size: String::new(),
            ext: FileExtension::JPG,
            url: None,
        }
    }
}

#[derive(
    Debug, serde::Serialize, serde::Deserialize, Clone, JsonSchema, PartialEq, Eq, Translate,
)]
pub enum FileExtension {
    #[serde(alias = "jpg", alias = "jpeg", alias = "JPEG")]
    #[translate(ko = "JPG", en = "JPG")]
    JPG = 1,

    #[serde(alias = "png", alias = "PNG")]
    #[translate(ko = "PNG", en = "PNG")]
    PNG = 2,

    #[serde(alias = "pdf", alias = "PDF")]
    #[translate(ko = "PDF", en = "PDF")]
    PDF = 3,

    #[serde(alias = "zip", alias = "ZIP")]
    #[translate(ko = "ZIP", en = "ZIP")]
    ZIP = 4,

    #[serde(alias = "doc", alias = "docx", alias = "word", alias = "WORD")]
    #[translate(ko = "WORD", en = "WORD")]
    WORD = 5,

    #[serde(alias = "ppt", alias = "pptx", alias = "PPTX")]
    #[translate(ko = "PPTX", en = "PPTX")]
    PPTX = 6,

    #[serde(alias = "xls", alias = "xlsx", alias = "excel", alias = "EXCEL")]
    #[translate(ko = "EXCEL", en = "EXCEL")]
    EXCEL = 7,

    #[serde(alias = "mp4", alias = "MP4")]
    #[translate(ko = "MP4", en = "MP4")]
    MP4 = 8,

    #[serde(alias = "mov", alias = "MOV")]
    #[translate(ko = "MOV", en = "MOV")]
    MOV = 9,

    #[serde(alias = "mkv", alias = "MKV")]
    #[translate(ko = "MKV", en = "MKV")]
    MKV = 10,
}

File: ts-packages/web/src/features/spaces/files/types/file.ts
interface FileModel {
  id: string;
  name: string;
  size: string;
  ext: FileExtension;
  url?: string | null;
}
export default FileModel;

export enum FileExtension {
  None = 'none',
  PNG = 'png',
  JPG = 'jpg',
  GIF = 'gif',
  WEBM = 'webm',
  SVG = 'svg',
  AI = 'ai',
  ZIP = 'zip',
  PDF = 'pdf',
  XLSX = 'xlsx',

  // 3D Model
  GLB = 'glb',
  GLTF = 'gltf',
  WORD = 'word',
  EXCEL = 'excel',
  // Audio
  MP3 = 'mp3',
  WAV = 'wav',

  // Video
  MP4 = 'mp4',
  MOV = 'mov',
  MKV = 'mkv',

  // Etc
  PPTX = 'pptx',
}

export function toFileExtension(
  input: string | undefined | null,
): FileExtension {
  if (!input) return FileExtension.PDF;

  const s = input.trim().toLowerCase();

  const mimeMap: Record<string, string> = {
    'image/jpeg': 'jpg',
    'image/jpg': 'jpg',
    'image/png': 'png',
    'application/pdf': 'pdf',
    'application/zip': 'zip',
    'application/x-zip-compressed': 'zip',
    'application/msword': 'doc',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document':
      'docx',
    'application/vnd.ms-powerpoint': 'ppt',
    'application/vnd.openxmlformats-officedocument.presentationml.presentation':
      'pptx',
    'application/vnd.ms-excel': 'xls',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': 'xlsx',
    'text/csv': 'csv',
    'video/mp4': 'mp4',
    'video/mkv': 'mkv',
    'video/quicktime': 'mov',
  };

  let ext = mimeMap[s];

  if (!ext) {
    if (s.includes('.')) {
      ext = s.split('.').pop() || '';
    } else if (s.startsWith('.')) {
      ext = s.slice(1);
    } else {
      ext = s;
    }
  }

  switch (ext) {
    case 'jpg':
    case 'jpeg':
      return FileExtension.JPG;
    case 'png':
      return FileExtension.PNG;
    case 'pdf':
      return FileExtension.PDF;
    case 'zip':
      return FileExtension.ZIP;
    case 'doc':
    case 'docx':
      return FileExtension.WORD;
    case 'ppt':
    case 'pptx':
      return FileExtension.PPTX;
    case 'xls':
    case 'xlsx':
    case 'csv':
      return FileExtension.EXCEL;
    case 'mp4':
      return FileExtension.MP4;
    case 'mkv':
      return FileExtension.MKV;
    case 'mov':
      return FileExtension.MOV;
    default:
      return FileExtension.PDF; // 모르면 PDF로
  }
}

export type BackendFile = Omit<FileModel, 'ext'> & { ext: string };

export const toBackendFile = (f: FileModel): BackendFile => ({
  ...f,
  ext: f.ext.toUpperCase(),
});
